MakieCore/src/recipes.jl

```diff
@@ -189,6 +189,365 @@ macro recipe(theme_func, Tsym::Symbol, args::Symbol...)
+            thm = theme(scene)
+            dict = Dict{Symbol,Any}()
+            $(closure_exprs...)
+            return dict
+        end
+        DocumentedAttributes(d, closure)
+    end
+end
+
+
+macro recipe(Tsym::Symbol, args...)
+
+    funcname_sym = to_func_name(Tsym)
+    funcname! = esc(Symbol("$(funcname_sym)!"))
+    PlotType = esc(Tsym)
+    funcname = esc(funcname_sym)
+
+    syms = args[1:end-1]
+    for sym in syms
+        sym isa Symbol || throw(ArgumentError("Found argument that is not a symbol in the position where optional argument names should appear: $sym"))
+    end
+    attrblock = args[end]
+    if !(attrblock isa Expr && attrblock.head === :block)
+        throw(ArgumentError("Last argument is not a begin end block"))
+    end
+    # attrblock = expand_mixins(attrblock)
+    # attrs = [extract_attribute_metadata(arg) for arg in attrblock.args if !(arg isa LineNumberNode)]
+
+    docs_placeholder = gensym()
+
+    attr_placeholder = gensym()
+
+    q = quote
+        # This part is as far as I know the only way to modify the docstring on top of the
+        # recipe, so that we can offer the convenience of automatic augmented docstrings
+        # but combine them with the simplicity of using a normal docstring.
+        # The trick is to mark some variable (in this case a gensymmed placeholder) with the
+        # Core.@__doc__ macro, which causes this variable to get assigned the docstring on top
+        # of the @recipe invocation. From there, it can then be retrieved, modified, and later
+        # attached to plotting function by using @doc again. We also delete the binding to the
+        # temporary variable so no unnecessary docstrings stay in place.
+        Core.@__doc__ $(esc(docs_placeholder)) = nothing
+        binding = Docs.Binding(@__MODULE__, $(QuoteNode(docs_placeholder)))
+        user_docstring = if haskey(Docs.meta(@__MODULE__), binding)
+            _docstring = @doc($docs_placeholder)
+            delete!(Docs.meta(@__MODULE__), binding)
+            _docstring
+        else
+            "No docstring defined.\n"
+        end
+
+        const MakieCore.$(attr_placeholder) = @DocumentedAttributes $attrblock
+
+        $(funcname)() = not_implemented_for($funcname)
+        const $(PlotType){$(esc(:ArgType))} = Plot{$funcname,$(esc(:ArgType))}
+        $(MakieCore).plotsym(::Type{<:$(PlotType)}) = $(QuoteNode(Tsym))
+        function ($funcname)(args...; kw...)
+            kwdict = Dict{Symbol, Any}(kw)
+            _create_plot($funcname, kwdict, args...)
+        end
+        function ($funcname!)(args...; kw...)
+            kwdict = Dict{Symbol, Any}(kw)
+             _create_plot!($funcname, kwdict, args...)
+        end
+        function MakieCore.is_attribute(T::Type{<:$(PlotType)}, sym::Symbol)
+            sym in MakieCore.attribute_names(T)
+        end
+
+        function MakieCore.attribute_names(::Type{<:$(PlotType)})
+            keys(MakieCore.$(attr_placeholder).d)
+        end
```

Comment:
Is it possible to easily override this or add new attributes, when e.g. editing the recipe / adding new rendering methods?

Reply:
Not sure what you mean, when would you want to do this? You can always reevaluate the edited recipe when you want to make changes

Reply:
I was specifically looking at adding another keyword to control `poly` rendering in CairoMakie, so was wondering if that could be picked up on only when CairoMakie is loaded, for example.  Or, if some extension from another package offers a rendering feature, it could hook in to say "I consume this keyword argument so mark it as valid" or something.

Reply:
No not currently. When discussing backend attributes with Simon we decided it's usually better to go via types for special backend functionality, and that all available attributes should ideally work across all backends. Even if that's not achievable it should always be the goal, so it shouldn't be easy to tack on keywords.